cpp/src/arrow/tensor.cc

```diff
@@ -127,14 +163,25 @@ Status CheckTensorStridesValidity(const std::shared_ptr<Buffer>& data,
     return Status::OK();
   }
 
-  std::vector<int64_t> last_index(shape);
-  const int64_t n = static_cast<int64_t>(shape.size());
-  for (int64_t i = 0; i < n; ++i) {
-    --last_index[i];
+  // Check the largest offset can be computed without overflow
+  const auto ndim = shape.size();
+  int64_t largest_offset = 0;
+  for (auto i = decltype(ndim){0}; i < ndim; ++i) {
+    if (strides[i] <= 0) continue;
```

Comment:
Are negative strides possible? If so, we shouldn't ignore them.

Reply:
Oops. I should check `dim_offset < 0` after `MultiplyWithOverflow` instead of `strides` here. I'll fix it.

Reply:
I noticed that we don't accept Numpy's arrays that have negative strides in `NdarrayToTensor()`: https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/numpy_convert.cc#L228-L230

But as `Tensor::Make()` and `Tensor::Tensor()` haven't rejected negative strides so far, I'd like to take care of negative strides here.